### PR TITLE
Support .well-known delegation when issuing certificates through ACME

### LINKED ISSUE #4552: ACME should support being given an alternate hostname
For people who don't want to deal with certificate management and also don't have their synapse and webserver in the same place. 

ie: support .well-known delegation from 1708 and ACME.

### PR DESCRIPTION
**Pull request by babolivier**
Fixes #4552

### CODE REVIEW

**Comment by hawkowl** [comment]
I feel like this should be manually configured, not done automagically?

**Comment by babolivier** [comment]
Hmm, I'd figure it would, especially since the original issue was mentioning "support .well-known delegation from 1708 and ACME", which I interpret as "synapse checking whether there's a .well-known in order to know which domain it should issue a cert for".

File: changelog.d/4652.feature [hunk 1]
```diff
@@ -0,0 +1 @@
+Support .well-known delegation when issuing certificates through ACME
```
Comment (anoadragon453) [inline changelog.d/4652.feature]: Changelog entries should have punctuation :)

```suggestion
Support .well-known delegation when issuing certificates through ACME.
```

File: synapse/handlers/acme.py [hunk 2]
```diff
@@ -123,15 +125,34 @@ def start_listening(self):
...
+            reactor=self.reactor,
+        )
+        delegated = yield agent._get_well_known(bytes(self.hs.hostname, "ascii"))
+
+        # If .well-known is in use, use the delegated hostname instead of the
+        # homeserver's server_name.
+        if delegated:
+            cert_name = delegated.decode("ascii")
+            logger.info(
+                ".well-known is in use, provisionning %s instead of %s",
```
Comment (anoadragon453) [inline synapse/handlers/acme.py]:
```suggestion
                ".well-known is in use, provisioning %s instead of %s",
```

**Comment by richvdh** [comment]
I'm with hawkowl. It should be explicit. The .well-known check, if we do one at all, should just be a sanity-check, but I think it's overkill.

**Comment by erikjohnston** [comment]
FWIW I think it should be a config option too, otherwise weird things will happen if we start it up while `.well-known` is down (e.g. due to restart server at the same time). In general its good if you can always bring up a service, even when dependencies might be temporary down, otherwise restarting the service becomes a lot scarier as you don't know if it'll actually come back.

**Comment by babolivier** [comment]
Roger that, making it a config parameter. Following @erikjohnston's comment I'll remove the well-known check entirely.

File: synapse/handlers/acme.py [hunk 3]
```diff
@@ -123,15 +123,15 @@ def start_listening(self):
     @defer.inlineCallbacks
     def provision_certificate(self):
 
-        logger.warning("Reprovisioning %s", self.hs.hostname)
+        logger.warning("Reprovisioning %s", self.hs.config.acme_domain)
```
Comment (richvdh) [inline synapse/handlers/acme.py]: could you fetch `self.hs.config.acme_domain` into an instance member (`self._acme_domain`) in the constructor? That will mean (a) it'll be noticed sooner if there's a problem reading it; (b) it's more concise)

File: synapse/config/tls.py [hunk 4]
```diff
@@ -229,6 +230,12 @@ def default_config(self, config_dir_path, server_name, **kwargs):
             #
             # reprovision_threshold: 30
 
+            # What domain the certificate should be for. Only useful if
```
Comment (richvdh) [inline synapse/config/tls.py]: I think we can be clearer here.

> The domain that the certificate should be for. Normally this should be the same as your matrix domain (ie, 'server_name'), but, by putting a file at 'https://<server_name>/.well-known/matrix/server', you can delegate incoming traffic to another server. If you do that, you should give the target of the delegation here.
>
> For example: if your 'server_name' is 'example.com', but 'https://example.com/.well-known/matrix/server' delegates to 'matrix.example.com', you should put 'matrix.example.com' here.

Comment (babolivier) [reply synapse/config/tls.py]: Thanks. I wasn't happy about the documentation I had, but wasn't sure about how to improve it.

**Pull request merged**